internal/bus/bus_redis.go

```diff
@@ -19,71 +19,290 @@ import (
+	}
+
+	subList, ok := subLists[channel]
+	if !ok {
+		return
+	}
+	i := slices.Index(subList.subs, msgChan)
+	if i == -1 {
+		return
+	}
+
+	subList.subs = slices.Delete(subList.subs, i, i+1)
+	close(msgChan)
+
+	if len(subList.subs) == 0 {
+		delete(subLists, channel)
+		r.reconcileSubscriptions(channel)
+	}
+}
+
+func (r *redisMessageBus) readWorker() {
+	for {
+		msg, err := r.ps.ReceiveMessage(r.ctx)
+		if err != nil {
+			return
+		}
+
+		r.mu.Lock()
+		if subList, ok := r.subs[msg.Channel]; ok {
+			subList.publish(msg)
+		}
+		if subList, ok := r.queues[msg.Channel]; ok {
+			subList.publishQueue(msg)
+		}
+		r.mu.Unlock()
+	}
+}
+
+func (r *redisMessageBus) reconcileSubscriptions(channel string) {
+	r.dirtyChannels.PushBack(channel)
+	r.enqueueWriteOp(&redisReconcileSubscriptionsOp{r})
+}
+
+func (r *redisMessageBus) enqueueWriteOp(op redisWriteOp) {
+	r.ops.PushBack(op)
+	select {
+	case r.wakeup <- struct{}{}:
+	default:
+	}
+}
+
+func (r *redisMessageBus) writeWorker() {
+	for range r.wakeup {
+		r.mu.Lock()
+		for r.ops.Len() > 0 {
+			op := r.ops.PopFront()
+			r.mu.Unlock()
+			op.run()
+			r.mu.Lock()
+		}
+		r.mu.Unlock()
+	}
+}
+
+type redisWriteOp interface {
+	run()
+}
+
+type redisPublishOp struct {
+	*redisMessageBus
+	channel string
+	message []byte
+}
+
+func (r *redisPublishOp) run() {
+	r.rc.Publish(r.ctx, r.channel, r.message)
+}
+
+type redisReconcileSubscriptionsOp struct {
+	*redisMessageBus
+}
+
+func (r *redisReconcileSubscriptionsOp) run() {
+	r.mu.Lock()
+	for r.dirtyChannels.Len() > 0 {
+		subscribe := map[string]struct{}{}
+		unsubscribe := map[string]struct{}{}
+		for r.dirtyChannels.Len() > 0 {
+			c := r.dirtyChannels.PopFront()
+			_, current := r.currentChannels[c]
+			desired := r.subs[c] != nil || r.queues[c] != nil
+			if !current && desired {
+				delete(unsubscribe, c)
+				subscribe[c] = struct{}{}
+			} else if current && !desired {
+				delete(subscribe, c)
+				unsubscribe[c] = struct{}{}
+			}
+		}
+		r.mu.Unlock()
+
+		var subscribeErr, unsubscribeErr error
+		if len(subscribe) != 0 {
+			subscribeErr = r.ps.Subscribe(r.ctx, maps.Keys(subscribe)...)
+		}
+		if len(unsubscribe) != 0 {
+			unsubscribeErr = r.ps.Unsubscribe(r.ctx, maps.Keys(unsubscribe)...)
+		}
+
+		if err := multierr.Combine(subscribeErr, unsubscribeErr); err != nil {
+			logger.Error(err, "redis subscription reconciliation failed")
+			time.Sleep(reconcilerRetryInterval)
+		}
+
+		r.mu.Lock()
+		if subscribeErr != nil {
+			for c := range subscribe {
+				r.dirtyChannels.PushFront(c)
+			}
+		} else {
+			for c := range subscribe {
+				r.currentChannels[c] = struct{}{}
+			}
+		}
+		if unsubscribeErr != nil {
+			for c := range unsubscribe {
+				r.dirtyChannels.PushFront(c)
+			}
+		} else {
+			for c := range unsubscribe {
+				delete(r.currentChannels, c)
+			}
+		}
+	}
+	r.mu.Unlock()
+}
+
+type redisSubList struct {
+	subs []chan *redis.Message
+	next int
+}
+
+func (r *redisSubList) publishQueue(msg *redis.Message) {
```

Comment:
nit: should these be called dispatchQueue and dispatch vs publish? When I read it initially, it sounded like publishing to Redis.

Reply:
yes, but i copied the name from the local implementation for consistency